# <font size = 4 color = black face = "Times New Roman">Individual Assignment Week 3
--------
<font size = 4 color = black face = "Times New Roman">By 黄家乐 518070910071
--------

In [1]:
import numpy as np
from sympy import *

#### <font size = 4 color = black face = "Times New Roman">Problem 1

<font size = 4 color = black face = "Times New Roman">Fom this problem, I wil write them as:

a) ${ln(x + 1) - ln(x) = ln(1+\frac{1}{x}) }$

b) ${\sqrt{x^2 +1}-x = {\frac {1} {\sqrt{x^2 + 1} + x}}}$

c) ${cos^2(x) - sin^2(x) = cos(2x)}$

d) ${\sqrt{\frac{1+cos(x)} {2}} = cos({\frac{x}{2}})}$

#### <font size = 4 color = black face = "Times New Roman">Problem 2

<font size = 4 color = black face = "Times New Roman">To solve this problem we need to know the double-precision of my computer. To make it more easier, I will use it from  my last homework, where $\epsilon = 1.0 \times 10^{-16}$

The formula what I use to solve the problem is 

a) $\epsilon_{tot}=\epsilon_{approx}+\epsilon_{ro}={\frac{1}{N^2}+\sqrt{N}\epsilon_{m}}$

b) $\epsilon_{tot}=\epsilon_{approx}+\epsilon_{ro}={\frac{2}{N^4}+\sqrt{N}\epsilon_{m}}$ 


<font size = 4 color = black face = "Times New Roman">My answers:

a) ${\frac{d\epsilon_{tot}}{dN}}=0 \Rightarrow N^{5/2} = {\frac{4}{10^{-16}}}\Rightarrow N = 4373448 \Rightarrow \epsilon_{tot} = 2.6 \times 10^{-13} $ 

In [2]:
eps_m = 10 ** (-16)
N = (4 / eps_m) ** (2 / 5)
print(N)

4373448.295773116


In [3]:
N = 4373448
eps_tot = 1 / ((N) ** 2) + np.sqrt(N) * eps_m
print(eps_tot)

2.6140988814781887e-13


b) ${\frac{d\epsilon_{tot}}{dN}}=0 \Rightarrow N^{9/2} = {\frac{16}{10^{-16}}}\Rightarrow N = 6655 \Rightarrow \epsilon_{tot} = 9.2 \times 10^{-15} $ 

In [4]:
eps_m = 10 ** (-16)
N = (16 / eps_m) ** (2 / 9)
print(N)

6654.842383978482


In [5]:
N = 6655
eps_tot = 2 / ((N) ** 4) + np.sqrt(N) * eps_m
print(eps_tot)

9.177436952478371e-15


#### <font size = 4 color = black face = "Times New Roman">Problem 3

<font size = 4 color = black face = "Times New Roman">We can think of this problem as solving a system of quadrilaterals. So the problem becomes more easier.

Here, we have:

${\pi_8 = \pi_{\infty} + \frac{c_1}{8} + \frac {c_2}{8^2} + \frac{c_3}{8^3}=3.061467}$

${\pi_{16} = \pi_{\infty} + \frac{c_1}{16} + \frac {c_2}{16^2} + \frac{c_3}{16^3}=3.121445}$

${\pi_{32} = \pi_{\infty} + \frac{c_1}{32} + \frac {c_2}{32^2} + \frac{c_3}{32^3}=3.136548}$

${\pi_{64} = \pi_{\infty} + \frac{c_1}{64} + \frac {c_2}{64^2} + \frac{c_3}{64^3}=3.140331}$

In [6]:
pi_inf = symbols('x')
c_1 = symbols('y')
c_2 = symbols('z')
c_3 = symbols('f')
pi_8 = pi_inf + c_1 / 8 + c_2 / (8 ** 2) + c_3 /(8 ** 3) - 3.061467
pi_16 = pi_inf + c_1 / 16 + c_2 / (16 ** 2) + c_3 /(16 ** 3) - 3.121445
pi_32 = pi_inf + c_1 / 32 + c_2 / (32 ** 2) + c_3 /(32 ** 3) - 3.136548
pi_64 = pi_inf + c_1 / 64 + c_2 / (64 ** 2) + c_3 /(64 ** 3) - 3.140331
solve([pi_8,pi_16,pi_32,pi_64],[pi_inf,c_1,c_2,c_3])

{x: 3.14158376190476,
 y: 0.00107733333333333,
 z: -5.20942933333333,
 f: 0.586703238095238}

From it, we get $\pi_{\infty}=3.14158376190476$, and obviously $c_1$ is most significant when $k$ is big enough. Because when $k$ is large, $|\frac{c_1}{k}|>>|\frac{c_2}{k^2}|>>|\frac{c_3}{k^3}|$.

#### <font size = 4 color = black face = "Times New Roman">Problem 4

By using my knowledge, I know that to make $g(x)$ approximate $\sqrt5$ I should choose $a=5$.

In [7]:
x = 2
for i in range (1,1000000):
    x = 1 / 2 * (x + 5 / x)
print(x)
print(np.sqrt(5))

2.23606797749979
2.23606797749979


<font size = 4 color = black face = "Times New Roman">The answer is right!